ext/dom/parentnode.c

```diff
@@ -350,19 +382,29 @@ void dom_parent_node_before(dom_object *context, zval *nodes, int nodesc)
 	newchild = fragment->children;
 
 	if (newchild) {
+		if (!parentNode->children) {
+			nextsib = NULL;
+		} else if (beforefirstchild) {
+			nextsib = parentNode->children == nextsib ? nextsib : afternextsib;
+		} else {
+			nextsib = parentNode->children == prevsib ? prevsib->next : nextsib;
```

Comment:
Question: as far as I can see, prevsib->next will always equal nextsib. So that means this check and assignment is redundant, no?

Reply:
```c
// ext/dom/parentnode.c:170
if (newNode->parent != NULL) {
	xmlUnlinkNode(newNode);
}
```
`xmlUnlinkNode` will detach the relation between with `parent node` and `child node`.


```c
if (!parentNode->children) {
    nextsib = NULL;
}
```
It mean `first node` and `child node` are both parameters to  `DOMElement::before()`  method.
So nextslib is null.

```c
if (beforefirstchild) {
	nextsib = parentNode->children == nextsib ? nextsib : afternextsib;
}
```
It mean that the `new node` will insert before `$doc->documentElement->firstChild`.
Let's discuss two possible scenarios.
1. `nextsib` does not change means that `$doc->documentElement->firstChild` is not the parameter to `DOMElement::before()`  method.
2. `nextsib = afternextsib` means that `$doc->documentElement->firstChild` is the parameter to `DOMElement::before()` method and `$doc->documentElement->firstChild` will removed from node tree.

It means that the `new node` will insert before `$doc->documentElement->lastChild` if neither of the above conditions is satisfied.
Let's discuss two possible scenarios too.
1. If `parentNode->children == prevsib` is true, it means that `$doc->documentElement->firstChild` is not the parameter to `DOMElement::before()`  method so `nextsib = prevsib->next`.
The reason we use `prevsib->next` is that `$doc->documentElement->lastChild` may be a parameter to `DOMElement::before()` and `$doc->documentElement->lastChild` will removed from node tree.
2. If `parentNode->children == prevsib` is false, it means that `$doc->documentElement->firstChild` is the parameter to `DOMElement::before()`  method so `nextsib` does not change.

Reply:
Okay that seems to make sense, thanks for explaining.